src/main/java/com/hivemq/extensions/handler/PubackInterceptorHandler.java

```diff
@@ -0,0 +1,441 @@
+        final SettableFuture<Void> interceptorFuture = SettableFuture.create();
+        final PubackInboundInterceptorContext interceptorContext = new PubackInboundInterceptorContext(
+                PubackInboundInterceptorTask.class, clientId, input, output, interceptorFuture,
+                pubackInboundInterceptors.size());
+
+        for (final PubackInboundInterceptor interceptor : pubackInboundInterceptors) {
+
+            if (interceptorFuture.isDone()) {
+                break;
+            }
+
+            final HiveMQExtension plugin = hiveMQExtensions.getExtensionForClassloader(
+                    (IsolatedPluginClassloader) interceptor.getClass().getClassLoader());
+
+            //disabled extension would be null
+            if (plugin == null) {
+                interceptorContext.increment();
+                continue;
+            }
+            final PubackInboundInterceptorTask interceptorTask =
+                    new PubackInboundInterceptorTask(interceptor, interceptorFuture, plugin.getId());
+
+            pluginTaskExecutorService.handlePluginInOutTaskExecution(
+                    interceptorContext, input, output, interceptorTask);
+        }
+
+        final InboundInterceptorFutureCallback callback =
+                new InboundInterceptorFutureCallback(output, puback, ctx);
+        Futures.addCallback(interceptorFuture, callback, ctx.executor());
+        return true;
+    }
+
+    private static class InboundInterceptorFutureCallback implements FutureCallback<Void> {
+
+        private final @NotNull PubackInboundOutputImpl output;
+        private final @NotNull PUBACK puback;
+        private final @NotNull ChannelHandlerContext ctx;
+
+        InboundInterceptorFutureCallback(
+                final @NotNull PubackInboundOutputImpl output,
+                final @NotNull PUBACK puback,
+                final @NotNull ChannelHandlerContext ctx) {
+            this.output = output;
+            this.puback = puback;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public void onSuccess(final @Nullable Void result) {
+            try {
+                final PUBACK finalPuback = PUBACK.createPubackFrom(output.getPubackPacket());
+                ctx.fireChannelRead(finalPuback);
+            } catch (final Exception e) {
+                log.error("Exception while modifying an intercepted PUBACK message.", e);
+                ctx.fireChannelRead(puback);
+            }
+        }
+
+        @Override
+        public void onFailure(final @NotNull Throwable t) {
+            //should never happen, since the settable future never sets an exception
+            ctx.channel().close();
+        }
+    }
+
+    private static class PubackInboundInterceptorContext extends PluginInOutTaskContext<PubackInboundOutputImpl> {
+
+        private final @NotNull PubackInboundInputImpl input;
+        private final @NotNull PubackInboundOutputImpl output;
```

Comment:
Why is the output stored here, it is provided to the pluginPost method, so this is unnecessary